Hello Erik!

My name is Olga. I'm happy to review your project today. 

When I will see mistake at the first time, I will just point it out. I let you find it and fix it by yourself. I'm trying to prepare you to work as an Data Analyst. SO, at a real job, your team lead will do the same. But if you can't handle this task yet, I will give you a more accurate hint at the next iteration.

You will find my comments below - **please do not move, modify or delete them**.

You can find my comments in green, yellow or red boxes like this:

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Success. Everything is done succesfully.

</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Remarks. Some recommendations.

</div>

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.

</div>

You can answer me by using this: 

<div class="alert alert-block alert-info">
<b>Student answer</b> <a class="tocSkip"></a>

</div>



<div class="alert alert-block alert-success">
<b>Overallv reviewer's comment</b> <a class="tocSkip"></a>
    
Eril, thank you for sending your project. You've done a really good job on it!

I like your filling missing numbers  and categorisation in your project. You did a great job! I am really impressed!

I'm glad to say that your project has been accepted. Keep at it. You're improving every day!
    
</div>

<div class="alert alert-block alert-warning">
<b>For your future projects</b> <a class="tocSkip"></a>
    
- Try to give tables meaningful names. Cars, clients, orders etc. data=data,df = DataFrame. Imagine, that you have 50 tables: df1, df2, df3, etc. it is hard to understand which table we need.
    
- It would be better to import all libraries in a first cell. It help us to set up of working environment.

- It is very good that you use Markdown! Please, use it more clearly. We can use #, ##, ###. etc for title, heading, subtitle, steps, etc. It helps to make our contents more structured
    
- I saw red automatic comments from Jupiter. We can remove them. We can add to libraries import:
  
        import warnings

        warnings.filterwarnings("ignore")

- We shoud use Kernel-Restart&Run All before sending project. It restart project and help to see errors.

- Analytical project should contains comments and conclusions during whole work. It make project completed and inforamble.

</div>




# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.


The purpose of this project is evaluate how the marital status and number of children affect the credit score of a potential customer applying for a loan. The hypothesis to be tested is that if a customer is single and has many children, then they are more likely to default on a loan.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Great that we have an introduction.
</div>

## Open the data file and have a look at the general information. 

In [1]:
# Loading all the libraries
import pandas as pd
import numpy as np

import nltk
from nltk.stem import WordNetLemmatizer
wordnet_lemma = WordNetLemmatizer()

# Load the data
data = pd.read_csv('/datasets/credit_scoring_eng.csv')

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

We should import libraries at the first cell and read files at the second. It would be easy to restart 1 cell than whole project, if we need to add library.
</div>

## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan


In [2]:
# Let's see how many rows and columns our dataset has

data

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

In [3]:
# let's print the first N rows

data.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

Through an initial overview, there is a value indicating 931 years employed and days employed should be positive instead of negative. Also, some entries under education indicate the same education but have varying capitilization. Although there aren't any missing values from this overview, it is necessary to look at/process more data.

In [4]:
# Get info on data
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


While there are 21525 total entries, a couple of columns have missing/null values. These columns are "days_employed" and "total_income".

In [5]:
# Let's look in the filtered table at the the first column with missing data
filtered_data = data[data['days_employed'].isna()]
filtered_data = filtered_data['days_employed']
filtered_data


12      NaN
26      NaN
29      NaN
41      NaN
55      NaN
         ..
21489   NaN
21495   NaN
21497   NaN
21502   NaN
21510   NaN
Name: days_employed, Length: 2174, dtype: float64

The missing values seem symmetric, as the total amount of 2174 missing values seem to be spread evenly at least throughout the top and bottom of the data. We are assuming from here that it is therefore uniformly spread within the dataset.

In [6]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.
null_data = data[data.isna().any(axis = 1)]
null_data

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

The number of rows in the null_data table match the number of missing values within both columns from the data, with a total 2174 rows. We can conclude from this that everywhere there all missing data for both "days_employed" and "total_income" occur on the same entry.

In [7]:
# calculate percentage of missing values compared to the whole dataset
percentage_null = len(null_data)/len(data) * 100
percentage_null

10.099883855981417

The percentage of null values is roughly 10%, which is a significant amount of missing data.

In [8]:
# Let's investigate clients who do not have data on identified characteristic and the column with the missing values
null_data[null_data.days_employed.isna()]['education_id']

12       1
26       1
29       1
41       1
55       1
        ..
21489    1
21495    1
21497    0
21502    1
21510    1
Name: education_id, Length: 2174, dtype: int64

In [9]:
# Checking distribution
null_data[null_data.days_employed.isna()]['education'].value_counts()


secondary education    1408
bachelor's degree       496
SECONDARY EDUCATION      67
Secondary Education      65
some college             55
Bachelor's Degree        25
BACHELOR'S DEGREE        23
primary education        19
Some College              7
SOME COLLEGE              7
Primary Education         1
PRIMARY EDUCATION         1
Name: education, dtype: int64

**Possible reasons for missing values in data**

After searching for dependencies throughout the other columns, there did not seem to be an obvious reason for the missing values.


In [10]:
# Checking the distribution in the whole dataset
data[data['days_employed'].isna()].sort_values(by='education')


children  days_employed  dob_years          education  education_id  \
10533         1            NaN         33  BACHELOR'S DEGREE             0   
2981          0            NaN         59  BACHELOR'S DEGREE             0   
10013         0            NaN         35  BACHELOR'S DEGREE             0   
21294         0            NaN         45  BACHELOR'S DEGREE             0   
3839          1            NaN         42  BACHELOR'S DEGREE             0   
...         ...            ...        ...                ...           ...   
4527          0            NaN         52       some college             2   
2171          0            NaN         31       some college             2   
9544          1            NaN         33       some college             2   
18962         0            NaN         26       some college             2   
13133         0            NaN         31       some college             2   

           family_status  family_status_id gender    income_type  debt  \
10533            married                 0      F       business     0   
2981             married                 0      M  civil servant     0   
10013            married                 0      F       business     1   
21294            married                 0      F       employee     0   
3839   civil partnership                 1      F       business     0   
...                  ...               ...    ...            ...   ...   
4527           unmarried                 4      F       employee     0   
2171             married                 0      M  civil servant     0   
9544           unmarried                 4      F       employee     0   
18962          unmarried                 4      M       employee     0   
13133          unmarried                 4      M       business     0   

       total_income                     purpose  
10533           NaN  buy commercial real estate  
2981            NaN    second-hand car purchase  
10013           NaN  buy commercial real estate  
21294           NaN        university education  
3839            NaN                         car  
...             ...                         ...  
4527            NaN         building a property  
2171            NaN        housing transactions  
9544            NaN       purchase of the house  
18962           NaN    real estate transactions  
13133           NaN                to own a car  

[2174 rows x 12 columns]

The distributions are identical in both the dataset and the filtered table.

**Conclusions**

There did not seem to be a pattern with the missing values within 'days_employed' and 'total_income'. In order to further evaluate the dataset however, it is best to correct any problematic data, including missing values, duplicates, discrepencies in string capitilizations, and values that don't make sense according to the category they are placed in.

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Also we have missing values in a total income and we need to check it. And if we use column with education, we need to use str.lower before.
</div>

## Data transformation



In [11]:
# Let's see all values in education column to check if and what spellings will need to be fixed
data['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [12]:
# Fix the registers if required
data['education'] = data['education'].str.lower()

In [13]:
# Checking all the values in the column to make sure we fixed them
data['education'].value_counts()


secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

[Check the data the `children` column]

In [14]:
# Let's see the distribution of values in the `children` column
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Of course, negative people cannot exist according to our modern understanding of mathematics and science. We will assume here that the 46 values of -1 should instead indicate 1 child.

Let's also assume that it is highly improbable that 76 people have exactly 20 children, and that those people actually intended to input 2 children.

The percentage of data changed within this column is:

In [15]:
percentage_children_to_change = (47 + 76)*100/data['children'].sum()
percentage_children_to_change

1.0603448275862069

1.2%

In [16]:
# [fix the data based on your decision]
data['children'] = data['children'].replace(-1, 1)
data['children'] = data['children'].replace(20, 2)

In [17]:
# Checking the `children` column again to make sure it's all fixed
data['children'].value_counts()


0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Well done!
</div>

We will assume that any value over 36500 days (100 years) is not accurate, and that any negative value should be positive.

In [18]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage
high_data = data[data['days_employed'] > 36500]
percentage_high_data = len(high_data)/len(data) * 100
percentage_high_data


16.004645760743323

In [19]:
high_data[high_data['days_employed'] > 36500]

children  days_employed  dob_years            education  education_id  \
4             0  340266.072047         53  secondary education             1   
18            0  400281.136913         53  secondary education             1   
24            1  338551.952911         57  secondary education             1   
25            0  363548.489348         67  secondary education             1   
30            1  335581.668515         62  secondary education             1   
...         ...            ...        ...                  ...           ...   
21505         0  338904.866406         53  secondary education             1   
21508         0  386497.714078         62  secondary education             1   
21509         0  362161.054124         59    bachelor's degree             0   
21518         0  373995.710838         59  secondary education             1   
21521         0  343937.404131         67  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
4      civil partnership                 1      F     retiree     0   
18       widow / widower                 2      F     retiree     0   
24             unmarried                 4      F     retiree     0   
25               married                 0      M     retiree     0   
30               married                 0      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21505  civil partnership                 1      M     retiree     0   
21508            married                 0      M     retiree     0   
21509            married                 0      M     retiree     0   
21518            married                 0      F     retiree     0   
21521            married                 0      F     retiree     0   

       total_income                                   purpose  
4         25378.572                         to have a wedding  
18         9091.804                  buying a second-hand car  
24        46487.558  transactions with commercial real estate  
25         8818.041                           buy real estate  
30        27432.971  transactions with commercial real estate  
...             ...                                       ...  
21505     12070.399                         to have a wedding  
21508     11622.175                                  property  
21509     11684.650                  real estate transactions  
21518     24618.344                         purchase of a car  
21521     24959.969                         purchase of a car  

[3445 rows x 12 columns]

In [20]:
negative_data = data[data['days_employed'] < 0]
negative_data_percentage = len(negative_data)/len(data)* 100
negative_data_percentage

73.89547038327527

# Address the problematic values, if they exist
16% of our data within the "days_employed" category reflect employment times of over 100 years, and in many cases well over. When attempting to convert these values to hours, the conversions seemed to represent an accurate depiction of the amount of days employed, especially considered that that the higher values correlated with the higher age.

Most of the data is also negative, and should be changed to positive days employed.

In [21]:
high_data.index

Int64Index([    4,    18,    24,    25,    30,    35,    50,    56,    71,
               78,
            ...
            21467, 21475, 21485, 21501, 21504, 21505, 21508, 21509, 21518,
            21521],
           dtype='int64', length=3445)

In [22]:
data.days_employed[high_data.index] = data.days_employed[high_data.index]/24

/tmp/ipykernel_45/644695215.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.days_employed[high_data.index] = data.days_employed[high_data.index]/24


In [23]:
# Check the result - make sure it's fixed
data['days_employed'] = data['days_employed'].abs()
data

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0   14177.753002         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0   14330.725172         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Correct!
    
</div>

In [24]:
# Check the `dob_years` for suspicious values and count the percentage
data[data['dob_years'] == 0]
percentage_zero_dob = len(data[data['dob_years'] == 0])/len(data) * 100
percentage_zero_dob

0.4692218350754936

There are 101 entries where the "dob_years" is equal to 0. We will replace these entries with the mean from the column since there does not seem to be a correlation with other columns.

# Address the issues in the `dob_years` column, if they exist


In [25]:
# Check the result - make sure it's fixed
data['dob_years'] = data['dob_years'].replace(0, data['dob_years'].mean())
len(data[data['dob_years'] == 0])

0

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

We can fill missing values by median or use days employed. :)
    
</div>

# Address the problematic values in `family_status`, if they exist


In [26]:
# Let's see the values for the column
data['family_status'].unique()


array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

'famiily_status' has no problematic values.

In [27]:
# Let's see the values in the column
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

'gender' has no problematic values.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Yep!
</div>

# Address the problematic values, if they exist


In [28]:
# Let's see the values in the column
data['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

# Address the problematic values, if they exist
'income_type' has no problematic values.

In [29]:
# Checking duplicates
data.duplicated().sum()

71

In [30]:
# Address the duplicates, if they exist
data[data.duplicated()].describe()
data = data.drop_duplicates()
data

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028       42.0    bachelor's degree             0   
1             1    4024.803754       36.0  secondary education             1   
2             0    5623.422610       33.0  secondary education             1   
3             3    4124.747207       32.0  secondary education             1   
4             0   14177.753002       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663       43.0  secondary education             1   
21521         0   14330.725172       67.0  secondary education             1   
21522         1    2113.346888       38.0  secondary education             1   
21523         3    3112.481705       38.0  secondary education             1   
21524         2    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21454 rows x 12 columns]

In [31]:
# Last check whether we have any duplicates
data.duplicated().sum()

0

In [32]:
# Check the size of the dataset that you now have after your first manipulations with it
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21454 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21454 non-null  float64
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21454 non-null  object 
dtypes: float64(3), int64(4), object(5)
memory usage: 2.1+ MB


The new dataset has some changes. First, the education category has been modified so that they are not case sensitive. Then, the entries where children were of -1 value were modified to be of +1 value, and those of 20 were grouped with those with 2 childred. Afterwards, it was found that roughly 73% of the data entries within 'days_employed') were of negative value, and these were all changed to be of absolute value. For days employed values that were greater than 36,500 (100 years), these were modified by dividing by 24 with the assumption that these are instead inputs of hours. For 'dob_years', there were 101 entries of 0 (about 0.5%). These were all replaced with the mean of that column. Finally, 71 duplicates were dropped. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Excellent! 
</div>

# Working with missing values

Dictionaries can help classify the data in useful ways, where a key (or the variable that we are testing our hypothesis on) can describe a value.

# Find the dictionaries


### Restoring missing values in `total_income`

The columns "days_employed" and "total_income" have missing values, each on the same indeces. We will fill these in with either a median or mean based on other characteristics within the dataset.

In [33]:
# Let's write a function that calculates the age category
def age_group(years):
    if 0 <= years < 10:
        return '0-9'
    elif 10 <= years < 20:
        return '10-19'
    elif 20 <= years < 30:
        return '20-29'
    elif 30 <= years < 40:
        return '30-39'
    elif 40 <= years < 50:
        return '40-49'
    elif 50 <= years < 60:
        return '50-59'
    elif 60 <= years < 70:
        return '60-69'
    elif 70 <= years < 80:
        return '70-79'
    elif 80 <= years < 90:
        return '80-89'
    elif 90 <= years < 100:
        return '90-99'




        
            
    

In [34]:
# Creating new column based on function
data['age_group'] = data['dob_years'].apply(age_group)


/tmp/ipykernel_45/3947402897.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['age_group'] = data['dob_years'].apply(age_group)


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Great work with age!
    
</div>

In [35]:
# Checking how values in the new column

data[['dob_years', 'age_group']]

dob_years age_group
0           42.0     40-49
1           36.0     30-39
2           33.0     30-39
3           32.0     30-39
4           53.0     50-59
...          ...       ...
21520       43.0     40-49
21521       67.0     60-69
21522       38.0     30-39
21523       38.0     30-39
21524       40.0     40-49

[21454 rows x 2 columns]

In [36]:
# Create a table without missing values and print a few of its rows to make sure it looks fine
filled_data = data.dropna()
filled_data

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028       42.0    bachelor's degree             0   
1             1    4024.803754       36.0  secondary education             1   
2             0    5623.422610       33.0  secondary education             1   
3             3    4124.747207       32.0  secondary education             1   
4             0   14177.753002       53.0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663       43.0  secondary education             1   
21521         0   14330.725172       67.0  secondary education             1   
21522         1    2113.346888       38.0  secondary education             1   
21523         3    3112.481705       38.0  secondary education             1   
21524         2    1984.507589       40.0  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose age_group  
0         40620.102    purchase of the house     40-49  
1         17932.802             car purchase     30-39  
2         23341.752    purchase of the house     30-39  
3         42820.568  supplementary education     30-39  
4         25378.572        to have a wedding     50-59  
...             ...                      ...       ...  
21520     35966.698     housing transactions     40-49  
21521     24959.969        purchase of a car     60-69  
21522     14347.610                 property     30-39  
21523     39054.888        buying my own car     30-39  
21524     13127.587             to buy a car     40-49  

[19351 rows x 13 columns]

In [37]:
# Look at the mean values for income based on your identified factors
mean_income=data.groupby('age_group')['total_income'].mean()
mean_income


age_group
10-19    16993.942462
20-29    25572.630177
30-39    28312.479963
40-49    28491.929026
50-59    25811.700327
60-69    23242.812818
70-79    20125.658331
Name: total_income, dtype: float64

In [38]:
# Look at the median values for income based on your identified factors
median_income = data.groupby('age_group')['total_income'].median()
median_income

age_group
10-19    14934.9010
20-29    22799.2580
30-39    24667.5280
40-49    24755.6960
50-59    22203.0745
60-69    19817.4400
70-79    18751.3240
Name: total_income, dtype: float64

In [39]:
mean_income=data.groupby('education')['total_income'].mean()
mean_income


education
bachelor's degree      33142.802434
graduate degree        27960.024667
primary education      21144.882211
secondary education    24594.503037
some college           29045.443644
Name: total_income, dtype: float64

In [40]:
median_income = data.groupby('education')['total_income'].median()
median_income

education
bachelor's degree      28054.5310
graduate degree        25161.5835
primary education      18741.9760
secondary education    21836.5830
some college           25618.4640
Name: total_income, dtype: float64

In [41]:
mean_income=data.groupby('children')['total_income'].mean()
mean_income

children
0    26422.404866
1    27368.627863
2    27478.854282
3    29322.623993
4    27289.829647
5    27268.847250
Name: total_income, dtype: float64

In [42]:
median_income = data.groupby('children')['total_income'].median()
median_income

children
0    23029.9535
1    23660.5630
2    23136.1155
3    25155.4480
4    24981.6340
5    29816.2255
Name: total_income, dtype: float64

In [43]:
mean_income=data.groupby('family_status')['total_income'].mean()
mean_income

family_status
civil partnership    26694.428597
divorced             27189.354550
married              27041.784689
unmarried            26934.069805
widow / widower      22984.208556
Name: total_income, dtype: float64

In [44]:
median_income = data.groupby('family_status')['total_income'].median()
median_income

family_status
civil partnership    23186.534
divorced             23515.096
married              23389.540
unmarried            23149.028
widow / widower      20514.190
Name: total_income, dtype: float64

It appears that education dfine income the most. We will use the median here to avoid outliers.

In [45]:
median_income = data.groupby('education')['total_income'].median()
median_income

education
bachelor's degree      28054.5310
graduate degree        25161.5835
primary education      18741.9760
secondary education    21836.5830
some college           25618.4640
Name: total_income, dtype: float64

In [46]:
#  Write a function that we will use for filling in missing values
median_income_dict=pd.Series(median_income).to_dict()
median_income_dict

{"bachelor's degree": 28054.531000000003,
 'graduate degree': 25161.5835,
 'primary education': 18741.976,
 'secondary education': 21836.583,
 'some college': 25618.464}

In [47]:
data['total_income']=data['total_income'].fillna(data.education.map(median_income_dict))

/tmp/ipykernel_45/3434248660.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['total_income']=data['total_income'].fillna(data.education.map(median_income_dict))


In [48]:
# Check if we got any errors
data.total_income.isna().sum()

0

In [49]:
# Checking the number of entries in the columns
data['total_income'].notnull().count()


21454

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Our salary is more depend on our profession. So, can you use another column for filling or use two.
    
</div>

###  Restoring values in `days_employed`

We will assume that the higher the age group, the higher the values for days employed

In [50]:
# Distribution of `days_employed` medians based on your identified parameters

median_days=data.groupby('age_group')['days_employed'].median()
median_days



age_group
10-19      724.492610
20-29     1005.629955
30-39     1601.784231
40-49     2108.529344
50-59     4796.767897
60-69    14801.234092
70-79    15055.708060
Name: days_employed, dtype: float64

In [51]:
# Distribution of `days_employed` means based on your identified parameters
mean_days=data.groupby('age_group')['days_employed'].mean()
mean_days

age_group
10-19      633.678086
20-29     1246.231977
30-39     2103.700591
40-49     3094.520150
50-59     7541.593571
60-69    12657.434080
70-79    13873.801149
Name: days_employed, dtype: float64

Since the 'days_employed' column had many high values that were converted, it is safer to use median here.

In [52]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter
median_days_dict=pd.Series(median_days).to_dict()
median_days_dict
data['days_employed']=data['days_employed'].fillna(data.age_group.map(median_days_dict))

/tmp/ipykernel_45/3324842953.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['days_employed']=data['days_employed'].fillna(data.age_group.map(median_days_dict))


In [53]:
# Check that the function works

data['days_employed'].iloc[12]

14801.234092431989

In [54]:
# Apply function to the income_type
median_days_dict=pd.Series(median_days).to_dict()
median_days_dict
data['days_employed']=data['days_employed'].fillna(data.age_group.map(median_days_dict))

/tmp/ipykernel_45/4253395483.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['days_employed']=data['days_employed'].fillna(data.age_group.map(median_days_dict))


In [55]:
# Check if function worked

data['days_employed'].isna().sum()

0

In [56]:
# Check the entries in all columns - make sure we fixed all missing values
data.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 21454 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     21454 non-null  float64
 2   dob_years         21454 non-null  float64
 3   education         21454 non-null  object 
 4   education_id      21454 non-null  int64  
 5   family_status     21454 non-null  object 
 6   family_status_id  21454 non-null  int64  
 7   gender            21454 non-null  object 
 8   income_type       21454 non-null  object 
 9   debt              21454 non-null  int64  
 10  total_income      21454 non-null  float64
 11  purpose           21454 non-null  object 
 12  age_group         21454 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 2.3+ MB


All columns are of the same value, with the correct size accounting for the removal of duplicates

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

For this column it is ok.    
</div>

## Categorization of data

In [57]:
# Print the values for your selected data for categorization

data['family_status'].value_counts()

married              12339
civil partnership     4151
unmarried             2810
divorced              1195
widow / widower        959
Name: family_status, dtype: int64

In [58]:
data['children'].value_counts()

0    14091
1     4855
2     2128
3      330
4       41
5        9
Name: children, dtype: int64

[Let's check unique values]

In [59]:
# Check the unique values
data['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [60]:
data = data.reset_index()

In [61]:
# Let's write a function to categorize the data based on common topics
purpose_list = []
for reason in data['purpose']:
    words = nltk.word_tokenize(reason)
    lemmas = [wordnet_lemma.lemmatize(w, pos = 'n') for w in words]
    lemmas = [l.lower() for l in lemmas]
    for i in lemmas:
        if i == 'house' or i == 'housing' or i =='estate' or i =='property':
            purpose_list.append('housing related')
        elif i == 'wedding':
            purpose_list.append('wedding related')
        elif i == 'car' or i == 'cars':
            purpose_list.append('car related')
        elif i == 'education' or i == 'going' or i == 'educated':
            purpose_list.append('education related')
        

len(purpose_list)

21454

In [62]:
data['purpose_category'] = pd.Series(data=purpose_list)
data['purpose_category']

0          housing related
1              car related
2          housing related
3        education related
4          wedding related
               ...        
21449      housing related
21450          car related
21451      housing related
21452          car related
21453          car related
Name: purpose_category, Length: 21454, dtype: object

In [63]:
data['purpose']

0          purchase of the house
1                   car purchase
2          purchase of the house
3        supplementary education
4              to have a wedding
                  ...           
21449       housing transactions
21450          purchase of a car
21451                   property
21452          buying my own car
21453               to buy a car
Name: purpose, Length: 21454, dtype: object

In [64]:
# Looking through all the numerical data in your selected column for categorization
data['total_income']

0        40620.102
1        17932.802
2        23341.752
3        42820.568
4        25378.572
           ...    
21449    35966.698
21450    24959.969
21451    14347.610
21452    39054.888
21453    13127.587
Name: total_income, Length: 21454, dtype: float64

In [65]:
# Getting summary statistics for the column
data['total_income'].describe()


count     21454.000000
mean      26466.156267
std       15701.145024
min        3306.762000
25%       17219.817250
50%       22583.630500
75%       31330.237250
max      362496.645000
Name: total_income, dtype: float64

We will use ranges according to the 25th, 50th, and 75th percentiles, as this will give us a generally uniform split of the data.

In [66]:
# Creating function for categorizing into different numerical groups based on ranges
def cat_income(total):
    if 3306 <= total < 17220:
        return 'within 25th percentile'
    elif 17220 <= total < 24797:
        return 'between 25th and 50th percentile'
    elif 24797 <= total < 31331:
        return 'between 50th and 75th percentile'
    elif total >= 31331:
        return 'between 75th and 100th percentile'


In [67]:
# Creating column with categories
data['income_category'] = data['total_income'].apply(cat_income)

In [68]:
# Count each categories values to see the distribution
data['income_category'].value_counts()

between 25th and 50th percentile     6770
between 75th and 100th percentile    5364
within 25th percentile               5364
between 50th and 75th percentile     3956
Name: income_category, dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Very good categorisation!</div>

## Checking the Hypotheses


**Is there a correlation between having children and paying back on time?**

In [69]:
# Check the children data and paying back on time
data[['children', 'debt']].value_counts()



children  debt
0         0       13028
1         0        4410
2         0        1926
0         1        1063
1         1         445
3         0         303
2         1         202
4         0          37
3         1          27
5         0           9
4         1           4
dtype: int64

In [70]:
# Calculating default-rate based on the number of children

row_values = [1063* 100/(1063+13028), 445* 100/(445+4410), 202 * 100/(202+1926), 27 * 100/(27+303), 4 * 100/(4+37), 0/5]
row_columns = ['zero', 'one', 'two', 'three', 'four', 'five']
child_rate = pd.Series(data=row_values, index=row_columns)
child_rate

zero     7.543822
one      9.165808
two      9.492481
three    8.181818
four     9.756098
five     0.000000
dtype: float64

**Conclusion**

When evaluating debts based on the number of children, it appears that those with at least one child tend to be more likely to default on a loan.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Intersting!</div>

**Is there a correlation between family status and paying back on time?**

In [71]:
# Check the family status data and paying back on time
data[['family_status','debt']].value_counts()





family_status      debt
married            0       11408
civil partnership  0        3763
unmarried          0        2536
divorced           0        1110
married            1         931
widow / widower    0         896
civil partnership  1         388
unmarried          1         274
divorced           1          85
widow / widower    1          63
dtype: int64

In [72]:
# Calculating default-rate based on family status

row_values = [931* 100/(931+11408), 388* 100/(388+3763), 274 * 100/(274+2536), 84 * 100/(84+1110), 63 * 100/(63+896)]
row_columns = ['married', 'civil partnership', 'unmarried', 'divorced', 'widow/widower']
family_rate = pd.Series(data=row_values, index=row_columns)
family_rate

married              7.545182
civil partnership    9.347145
unmarried            9.750890
divorced             7.035176
widow/widower        6.569343
dtype: float64

**Conclusion**



The highest default rate depending on family status is from the categories of civil partnership and unmarried, so there may be a correlation there.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Very good!</div>

**Is there a correlation between income level and paying back on time?**

In [73]:
# Check the income level data and paying back on time
data[['income_category', 'debt']].value_counts()






income_category                    debt
between 25th and 50th percentile   0       6170
between 75th and 100th percentile  0       4981
within 25th percentile             0       4937
between 50th and 75th percentile   0       3625
between 25th and 50th percentile   1        600
within 25th percentile             1        427
between 75th and 100th percentile  1        383
between 50th and 75th percentile   1        331
dtype: int64

In [74]:
# Calculating default-rate based on income level
row_values = [427* 100/(427+4937), 466* 100/(466+4898), 465 * 100/(465+4897),  383 * 100/(383+4981)]
row_columns = ['25th percentie', '50th percentile', '75th percentile', '100th percentile']
income_rate = pd.Series(data=row_values, index=row_columns)
income_rate

25th percentie      7.960477
50th percentile     8.687547
75th percentile     8.672137
100th percentile    7.140194
dtype: float64

**Conclusion**



Incomes between the 25th and 75th percentiles tend to default, but not by a large margin.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Correct!</div>

**How does credit purpose affect the default rate?**

In [75]:
# Check the percentages for default rate for each credit purpose and analyze them
data[['purpose_category', 'debt']].value_counts()
row_values = [782* 100/(782+10029), 403* 100/(403+3903), 370 * 100/(370+3643),  186 * 100/(186+2138)]
row_columns = ['housing', 'car', 'education', 'wedding']
purpose_rate = pd.Series(data=row_values, index=row_columns)
purpose_rate

housing      7.233373
car          9.359034
education    9.220035
wedding      8.003442
dtype: float64

**Conclusion**

The default rate based on credit purpose seems to be highest within the car and education categories.


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Yep!</div>

# General Conclusion 


The dataset required some changes in order to be evaluated effectively. Some entries in 'dob_years' were 0, which were replaced using the mean of the column. 71 duplicates were found and removed. The data in days_employed that were negative were changed to be positive, and data that reflected an impossible timeframe were assumed to be input as hours. This high data was modified and replaced accordingly. 

The columns 'days_employed' and 'total_income' had missing data, both of which were on the same indeces. In order to replace values in these columns, other characteristics within the data were used to see if there is a dependency. The data missing in 'days_employed' were filled with median values according to seperate age groups that were defined. There was a correlation of days employed with age group, with intuitively makes sense. The data missing in "total_income" were filled with median values for each education category. The higher the education, the higher the total income. 

After dealing with the artifacts in the data, the default rates depending on number of children, family status, income level, and credit purpose were evaluated. The parameters that seemed to predict the default rate the most were the family status, the number of children, and the credit purpose. As it pertains to the family status, it seems that those in a civil partnership or are unmarried tend to default at a higher rate than the other family statuses. In terms of children, the contrary of the hypotheses was true: Those who had at least once child were more likely to default when compared to those with no children. Lastly, the credit purposes concerning buying a car or paying for education had a higher default rate than those for weddings or housing. When looking at income level, there did not seem to be any correlation to be found.

Overall, as it relates to our original hypothesis, it was found that the family statuses of civil partnership and unmarried tend to default more often than others. Regarding children, it was found that those no children tend not to default on a loan, which was contrary to the initial assumption. In addition, the credit purposes of education and car purchases were found to also possibly correlate with a higher default rate.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Thank you for the details!</div>